concordium-consensus/src/Concordium/ImportExport.hs

```diff
@@ -150,90 +251,186 @@ exportDatabaseV3 ::
+                    Empty -> return (0, 1, Empty, Nothing)
+                    -- the following is safe due to normalization of the blockindex.
+                    blockIndexInit :|> (gd, chunks) -> case chunks of
+                        Empty -> fail $ "A section for genesis hash '" <> show gd <> "' in block index file has no entries."
+                        initChunks :|> BlockIndexChunkInfo{..} -> do
+                            -- the filename of the most recently exported chunk
+                            let lastChunkFile = outDir </> T.unpack filename
+                            -- if we overwrite files, back up the most recently exported
+                            -- chunk file so we can restore it if something goes wrong.
+                            when overwriteExistingChunks $ renameFile lastChunkFile $ lastChunkFile <> "~"
+                            -- evict the last chunk from the block index since we want to
+                            -- export blocks starting from the genesis index and the height
+                            -- of the first block contained within it
+                            return
+                                ( genesisIndex,
+                                  blockHeightFirst,
+                                  blockIndexInit :|> (gd, initChunks),
+                                  Just lastChunkFile
+                                )
             else do
                 createDirectoryIfMissing True outDir
-                -- we export blocks starting from height 1 because genesis blocks need not be exported
-                return (0, 1)
-    bracket
-        (openFile indexFile AppendMode)
-        hClose
-        (\indexHdl -> exportSections dbDir outDir indexHdl chunkSize genIndex startHeight)
-
--- |Export a database section for each genesis index.
+                -- we export blocks starting from height 1 because genesis blocks need not be exported.
+                return (0, 1, Empty, Nothing)
+
+    let lastChunkFileName = fromMaybe "" lastChunkFilenameM
+    let lastChunkBackupName = lastChunkFileName <> "~"
+    lastChunkBackupExists <- doesFileExist lastChunkBackupName
+
+    blockIndexTail <- exportSections dbDir outDir chunkSize overwriteExistingChunks genIndex startHeight blockIndex
+    case blockIndexTail of
+        -- in case nothing was was exported, restore the most recently
+        -- exported chunk s.t. the files and block index remain unchanged,
+        -- since this chunk was evicted from the index in the above.
+        Empty -> when lastChunkBackupExists $ renameFile lastChunkBackupName lastChunkFileName
+        -- otherwise write the updated block index file to the filesystem
+        -- and delete the backup file.
+        _ -> do
+            writeFile indexFile $ showBlockIndexFile $ normalizeBlockIndex $ blockIndex <> blockIndexTail
+            when lastChunkBackupExists $ removeFile lastChunkBackupName
```

Comment:
I guess technically these backup files could exist without this program creating them. So these checks have some caveats about them. Maybe mention that we assume that the tool is the only one writing to the supplied directory.